clients/src/main/java/org/apache/kafka/common/record/Records.java

```diff
@@ -70,6 +71,13 @@ public interface Records extends TransferableRecords {
      */
     AbstractIterator<? extends RecordBatch> batchIterator();
 
+    /**
+     * Return the last record batch if non-empty or an empty `Optional` otherwise.
+     *
+     * Note that this requires iterating over all the record batches and hence it's expensive.
+     */
+    Optional<RecordBatch> lastBatch();
```

Comment:
Do we need this as an interface level method? Asking because interface already exposes batchIterator(). Calculating lastBatch() is a helper function and we could just have it as a concrete implementation in AbstractRecords. Note that existing method firstBatch() is also not present in the interface.

Reply:
I don't see how it helps to only have it in `AbstractRecords`, the point is to encapsulate the implementation since the equivalent Java code is a lot more verbose than the Scala version.

Reply:
ok, I will leave it up to you.

In case you are curious, my thoughts were as follows but this is a trivial point and we can agree to disagree.

Do we anticipate concrete implementations of `Records` to provide an optimized implementation of `firstBatch()` and `lastBatch()` compared to scanning using batchIterator()? If yes, then they should be added to the contract. If we are only adding it here to hide the verbosity of same implementation that a callee will do, then should just use it as a helper (and not an interface methods). Helper methods go in abstract base class.

Reply:
`AbstractRecords` only exists because default methods weren't a thing then. Abstract classes that don't add any value like this one are a bit of anti-pattern.